master/internal/api_generic_tasks.go

```diff
@@ -433,12 +458,162 @@ func (a *apiServer) KillGenericTask(
+	if err != nil {
+		return nil, err
+	}
+	tasksToPause, err := a.GetTaskChildren(ctx, model.TaskID(req.TaskId), overrideStates)
+	if err != nil {
+		return nil, err
+	}
+	for _, childTask := range tasksToPause {
+		if childTask.NoPause == nil || !*childTask.NoPause {
+			allocationID, err := a.GetAllocationFromTaskID(ctx, childTask.TaskID)
+			if err != nil {
+				return nil, err
+			}
+			err = task.DefaultService.Signal(model.AllocationID(allocationID),
+				task.TerminateAllocation,
+				"user requested task kill")
+			if err != nil {
+				return nil, err
+			}
+			err = a.SetPausedState(ctx, childTask.TaskID)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	return &apiv1.PauseGenericTaskResponse{}, nil
+}
+
+func (a *apiServer) ResumeGenericTask(
+	ctx context.Context, req *apiv1.ResumeGenericTaskRequest,
+) (*apiv1.ResumeGenericTaskResponse, error) {
+	var taskModel model.Task
+	err := db.Bun().NewSelect().Model(&taskModel).Where("task_id = ?", req.TaskId).Scan(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// Validate state
+	if *taskModel.State != model.TaskStatePaused {
+		return nil, fmt.Errorf("cannot unpause task %s as it is not in paused state", req.TaskId)
+	}
+	var projectID int
+	if req.ProjectId != nil {
+		projectID = int(*req.ProjectId)
+	} else {
+		projectID = model.DefaultProjectID
+	}
+	// Tasks (and child tasks) that are killed, completed should not be paused
+	overrideStates := []model.TaskState{model.TaskStateCanceled, model.TaskStateCompleted}
+	if err != nil {
+		return nil, err
+	}
+	tasksToResume, err := a.GetTaskChildren(ctx, model.TaskID(req.TaskId), overrideStates)
+	if err != nil {
+		return nil, err
+	}
+	for _, childTask := range tasksToResume {
+		genericTaskSpec, _, _, err := a.getGenericTaskLaunchParameters(
+			ctx, nil, *childTask.Config, projectID,
+		)
+		if err != nil {
+			return nil, err
+		}
+		// check if job still in registry
+		_, exists := tasklist.GroupPriorityChangeRegistry.Load(*childTask.JobID)
+		if !exists {
+			priorityChange := func(priority int) error {
+				genericTaskSpec.GenericTaskConfig.Resources.SetPriority(&priority)
+				return nil
+			}
+			if err = tasklist.GroupPriorityChangeRegistry.Add(*childTask.JobID, priorityChange); err != nil {
+				return nil, err
+			}
+		}
+		allocationString, err := a.GetAllocationFromTaskID(ctx, childTask.TaskID)
+		if err != nil {
+			return nil, err
+		}
+		allocationID := model.AllocationID(allocationString)
+		logCtx := logger.Context{
+			"job-id":    childTask.JobID,
+			"task-id":   childTask.TaskID,
+			"task-type": model.TaskTypeGeneric,
+		}
+		onAllocationExit := func(ae *task.AllocationExited) {
+			syslog := logrus.WithField("component", "genericTask").WithFields(logCtx.Fields())
+			isPaused, err := a.m.db.IsPaused(ctx, childTask.TaskID)
+			if err != nil {
+				syslog.WithError(err).Error("on allocation exit")
+			}
+			if !isPaused {
```

Comment:
style nit again, flip the happy path to be unintended, special case to be intended